### ISSUE TITLE
Export key value/length does not work with Redis cluster.

### ISSUE THREAD

**Issue by hostmaster** [oliver006/redis_exporter#400]
**Describe the problem**
A node of a Redis cluster can redirect queries to the proper node. A cluster-aware Redis client is expected to connect to the specified node to retry a request.  `redis_exporter` does not support that. Export key value/length does not work with a cluster.

**What version of redis_exporter are you running?**
 - 1.9.0

**Running the exporter**
`./redis_exporter -redis.addr redis-cluster-endpoint:6379 -check-single-keys channel:updates`

```
INFO[0000] Redis Metrics Exporter v1.9.0    build date: 2020-07-07-02:25:01    sha1: 3e15a27ef8657e77100a55f10895e1a4fee70f79    Go: go1.14.4    GOOS: linux    GOARCH: amd64
INFO[0000] Providing metrics at :9121/metrics
ERRO[0022] MOVED 3527 10.0.1.100:6379
```

**Expected behavior**
Connect to the specified node to retry a request.

**Comment by oliver006** [comment]
Ah yes, it looks like the library that the exporter uses ([redigo](https://github.com/gomodule/redigo)) doesn't support cluster mode.

Changing the underlying Redis library is a good bit of work and there are currently no plans to do that.

**Comment by hostmaster** [comment]
That is entirely understandable. Would you mind mentioning the limitation in README? 

My golang skill is very basic however its seems like [redisc](https://github.com/mna/redisc) tries to keeps API compatible with [redigo](https://github.com/gomodule/redigo) where it is possible. It might help to reduce the scope of possible changes.

**Comment by oliver006** [comment]
Thanks for making me aware of redisc - didn't know about that.

I'm a little pressed for time right now but I can review and merge a PR for the README if you submit one.

**Comment by ashikjm** [comment]
@oliver006  Is this still open for picking up? I would like to work on this. Thanks.

**Comment by oliver006** [comment]
@ashikjm - yes, your help is very welcome!

Feel free to share an implementation plan if you want feedback before you dive in too deep but up to you.

**Comment by ashikjm** [comment]
I will do that. I will share what I am planning to do as soon as it is ready.

**Comment by ashikjm** [comment]
@oliver006 

What I am planning to do:
- Add an extra boolean flag 'isCluster'.
- Can take redisaddr as comma separated redis hosts in case of cluster (TBD)
- connectToRedis function will create connection objects based on the 'isCluster' flag
	- False condition will use the existing code
	- True:
		- create a cluster obj
		- create a connection obj to the cluster
		- will wrap this connection obj with retryconn method to do retries. This returns redis.conn so we don't need to change anything else.
		- return this obj.

I will start working on the PR and submit it as soon as it is ready.

**Comment by oliver006** [comment]
I prefer an approach that is transparent to the user and where the exporter / Prometheus server doesn't need to be aware of the cluster topology.

I think that would mean to use a library (that builds on redigo?) that supports cluster access (e.g. handles the redirects) and use that instead of redigo.

**Comment by ashikjm** [comment]
I tested **redisc** with a stand alone redis instance and it handles the commands fine. In that case, we can use redisc for connections to both (clusters and stand alone). In case it is a cluster, the only addition would be that the connection would be wrapped by a retryconn method but this would be transparent to the user. We will only ask user for an endpoint, determine whether it is a cluster or standalone etc within the connectToRedis function itself. Thoughts?

**Comment by oliver006** [comment]
I have not used/looked at redisc before but from a brief look at the description it sounds like this might work with (hopefully) minimal effort.

**Comment by ashikjm** [comment]
yes. The change should be minimal. I will raise a PR and discuss further there.

**Comment by chelomontilla** [comment]
Any updates on this @ashikjm ?

**Comment by EsmeAR** [comment]
@ashikjm also wanted to check if there was any news on this?

**Comment by ashikjm** [comment]
@chelomontilla @EsmeAR  The PR is under review. :)

**Comment by oliver006** [comment]
Closed in #531 thanks to @ashikjm
